pkg/csi/service/wcp/controller.go

```diff
@@ -1509,9 +1552,35 @@ func (c *controller) ControllerExpandVolume(ctx context.Context, req *csi.Contro
 		// If thr reqeust failed due to object not found, "csi.fault.NotFound" will be return.
 		// For all other cases, the faultType will be set to "csi.fault.Internal" for now.
 		// Later we may need to define different csi faults.
+		// Check if the volume contains CNS snapshots only for block volumes.
+		if cnsVolumeType == common.UnknownVolumeType {
+			cnsVolumeType, err = common.GetCnsVolumeType(ctx, c.manager, req.VolumeId)
+			if err != nil {
+				return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to determine CNS volume type for volume: %q. Error: %+v", req.VolumeId, err)
+			}
+			volumeType = convertCnsVolumeType(ctx, cnsVolumeType)
+		}
+		if cnsVolumeType == common.BlockVolumeType &&
+			commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot) {
+			snapshots, _, err := common.QueryVolumeSnapshotsByVolumeID(ctx, c.manager.VolumeManager, req.VolumeId,
+				common.QuerySnapshotLimit)
+			if err != nil {
+				return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to retrieve snapshots for volume: %s. Error: %+v", req.VolumeId, err)
+			}
+			if len(snapshots) == 0 {
+				log.Infof("no CNS snapshots found for volume: %s, the volume can be safely expanded",
+					req.VolumeId)
+			} else {
+				return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.FailedPrecondition,
+					"volume: %s with existing snapshots %v cannot be expanded, "+
+						"please delete snapshots before deleting the volume", req.VolumeId, snapshots)
+			}
 
+		}
 		isOnlineExpansionEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.OnlineVolumeExtend)
-		err := validateWCPControllerExpandVolumeRequest(ctx, req, c.manager, isOnlineExpansionEnabled)
+		err = validateWCPControllerExpandVolumeRequest(ctx, req, c.manager, isOnlineExpansionEnabled)
```

Comment:
Looks like we are missing a check for block volume type here as volume expansion is only supported for block volumes, but this is not related to your change.

Reply:
Let's not do bug fixes in feature changes. I'll speak to the QE about this, perhaps the checks are made elsewhere.